src/Tasks/GenerateResource.cs

```diff
@@ -2391,6 +2391,24 @@ internal bool StronglyTypedResourceSuccessfullyCreated
         private bool _useSourcePath = false;
 
 #endregion
+        internal void Setup()
+        {
+            static Assembly TryLoadAssembly(AssemblyName assemblyName)
+            {
+                // Look in the MSBuild folder for any unresolved reference. It may be a dependency
+                // of MSBuild or a task.
+                string msbuildDirectoryPath = Path.GetDirectoryName(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath);
+                string targetAssembly = Path.Combine(msbuildDirectoryPath, assemblyName.Name + ".dll");
+                if (File.Exists(targetAssembly))
+                {
+                    return Assembly.LoadFrom(targetAssembly);
```

Comment:
So when some dll isn't loaded, TryLoadAssembly runs which attempts to run this method. Do we need to worry about this throwing & catch it? Not too worried about that scenario, it looks like if this code is running then it would have already failed if we didn't do this. The error could be misleading though. Just thinking out loud.

From the docs [here](https://docs.microsoft.com/en-us/dotnet/standard/assembly/resolve-loads)

> When you write the handler, you should know which assemblies might cause the event to be raised. Your handler should return null for other assemblies.

Should we restrict the names of allowed assemblies that cause this event to run?

Reply:
This _should_ only be looking in the MSBuild folder, so it shouldn't have anything that has a .dll extension that isn't a real, loadable .dll, and if you added something badly formed and asked for it, then you're asking for trouble.

It's still possible, but the main reasons I'd prefer to avoid explicitly listing out those assemblies are that 1) It's messy, and 2) We don't actually know which assemblies we'll need in the future or to cover all code paths. I could just list out all the assemblies we currently ship, but if that changes, we'd have to go back and change the code in GenerateResource, too. Being open could give a misleading error as you say, but I'd prefer cleanliness and future-proof.

Reply:
WRT future debuggability I checked out the previous error/stack trace.
```
The "GenerateResource" task failed unexpectedly.
System.ArgumentException: Object type cannot be converted to target type.
   at Microsoft.Build.Tasks.ProcessResourceFiles.Run(TaskLoggingHelper log, ITaskItem[] assemblyFilesList, List`1 inputs, List`1 satelliteInputs, List`1 outputs, Boolean sourcePath, Boolean usePreserializedResources, String language, String namespacename, String resourcesNamespace, String filename, String classname, Boolean publicClass, Boolean extractingResWFiles, String resWOutputDirectory)
   at Microsoft.Build.Tasks.GenerateResource.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
Done executing task "GenerateResource" -- FAILED.
Done building target "CoreResGen" in project "ReproGenerateResourceErrorLibrary.csproj" -- FAILED.
 
```

Was it difficult to nail this issue down based on the error? If so, we should consider a try-catch around the call to `Run` catching the argument exception and specify some sort of "MSB1234: Hey this is an app domain issue".

Reply:
"Object type cannot be converted to target type" made no sense to me at first. It specified ProcessResourceFiles.Run, so I thought there would be a bug somewhere in Run, but it was the method call itself that failed, not anything within it, which was very confusing to me because I initially couldn't think of any way that was possible when it was ProcessResourceFiles.Run was defined in the same class/file as its caller.

The only difficulty is that Run does a _lot_ of work: reading the format, using that format to deserialize an object, reading another format, serializing that object again...I suspect there are several different ArgumentExceptions that could be thrown that have nothing to do with app domains.

That said, we could put MSB1234 looking for an ArgumentException _when the stack trace stops on ProcessResourceFiles.Run_. I don't think that can happen unless it's an app domain issue. Can you make sure we talk about this in PR review? I'm thinking it might be worth wrapping not just this but also other similar things in other tasks that use app domains with your MSB1234—it is likely that one or more of them are vulnerable to just this issue but haven't seen it yet. It may even be worth trying to make them all fail and seeing how many we can fix before customers see them. Though at that point, it's probably better to do it in a separate PR.